### DISCUSSION TITLE
Stability days for lock file maintenance

### DISCUSSION THREAD

**Discussion by stekern** [renovatebot/renovate#13460] [Request Help] [answered]
We'd like to keep our NPM lockfiles updated regularly, but it would be nice with a configurable grace period similar to `stabilityDays` for these kind of updates to avoid introducing potentially unstable (or malicious) changes.

Based on the docs, it does not seem like this type of functionality currently exists in Renovate.

- Is this a potential feature that could be looked at?
- Are there any current, alternative workflows for achieving something similar?

**Comment by rarkins** [comment, marked as answer]
This feature is not technically possible. Lock file maintenance is done by:
- Removing the lock file
- Running `npm install`

I recommend you pin all dependencies in your `package.json` unless you are publishing a library, and do not use version ranges.

**Reply by stekern** [reply]
Do you mean not technically possible given the current implementation of the lock file maintenance? Or not technically possible/feasible in the general sense (even as a new feature)?

I pin my dependencies to exact versions, but there's no guarantee that downstream packages are doing the same. In that case, lock file maintenance can potentially lead to update of an unstable or malicious package that's not a direct dependency in my project.

**Reply by viceice** [reply]
technically not possible by renovate, as lockfile change is done by the underlying native manager (eg. npm). So we can't do anything here at renovate.

**Reply by rarkins** [reply]
This is practically impossible regardless, unless you reimplement `npm`'s dependency resolving and completely reverse engineer the lockfile to generate it yourself.

@stekern were you expecting that Renovate can completely generate the lockfile itself, without using npm?

**Reply by stekern** [reply]
My naive thought was to have mechanism based on Renovate running `npm install`, doing a diff to check which packages were updated, fetching registry information for each of these and then only updating the ones that satisfy various conditions (e.g., stability days). I imagine something like this would not be very straight-forward to implement, especially if `npm` does not expose a flexible API for making granular changes to the lock files.

The interest in this type of functionality is motivated by recent issues with popular NPM packages (often nested deep in the dependency tree) being [hijacked](https://www.cisa.gov/uscert/ncas/current-activity/2021/10/22/malware-discovered-popular-npm-package-ua-parser-js), [intentionally corrupted](https://www.bleepingcomputer.com/news/security/dev-corrupts-npm-libs-colors-and-faker-breaking-thousands-of-apps/), etc. and that really highlight how vulnerable especially the Node.js ecosystem is.

I do think some more control of how lock files are updated would be a helpful addition, but perhaps this is a problem that should be fixed at the root by NPM rather than Renovate. I'm not sure.

**Reply by rarkins** [reply]
The challenge is that "small" changes of a single version of a dependency can have follow-on effects in the rest of the tree. So Renovate can't simply take what npm has done and then "adjust" a version here or there back. Dependencies can also "move" from being at top level to nested or vice versa, so even identifying what changed isn't trivial either.

In short, it would be better if there was a flag in npm which let you specify a duration/time period of e.g. 2 days unless manually overriden by some other CLI flag.
